subprojects/core/src/main/java/org/gradle/caching/internal/tasks/TaskOutputCacheCommandFactory.java

```diff
@@ -103,16 +125,45 @@ public BuildCacheKey getKey() {
             return new BuildCacheLoadCommand.Result<TaskOutputOriginMetadata>() {
                 @Override
                 public long getArtifactEntryCount() {
-                    return unpackResult.entries;
+                    return unpackResult.getEntries();
                 }
 
                 @Override
                 public TaskOutputOriginMetadata getMetadata() {
-                    return unpackResult.originMetadata;
+                    return unpackResult.getOriginMetadata();
                 }
             };
         }
 
+        private void updateSnapshots(ImmutableListMultimap<String, FileSnapshot> propertiesFileSnapshots) {
+            ImmutableMap<String, ResolvedTaskOutputFilePropertySpec> propertySpecsMap = Maps.uniqueIndex(outputProperties, new Function<TaskOutputFilePropertySpec, String>() {
+                @Override
+                public String apply(TaskOutputFilePropertySpec propertySpec) {
+                    return propertySpec.getPropertyName();
+                }
+            });
+            ImmutableSortedMap.Builder<String, FileCollectionSnapshot> propertySnapshotsBuilder = ImmutableSortedMap.naturalOrder();
+            for (String propertyName : propertiesFileSnapshots.keySet()) {
+                ResolvedTaskOutputFilePropertySpec property = propertySpecsMap.get(propertyName);
+                List<FileSnapshot> fileSnapshots = propertiesFileSnapshots.get(propertyName);
+                FileCollectionSnapshotBuilder builder = new FileCollectionSnapshotBuilder(UNORDERED, OutputPathNormalizationStrategy.getInstance(), stringInterner);
+                builder.visitFileTreeSnapshot(fileSnapshots);
```

Comment:
I've replaced the test with one that uses a couple custom tasks instead of Java tests.